runtime/visualize.cc

```diff
@@ -0,0 +1,559 @@
+#include "runtime/visualize.h"
+
+#include <algorithm>
+#include <atomic>
+#include <cassert>
+#include <cstddef>
+#include <cstdlib>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "runtime/buffer.h"
+#include "runtime/depends_on.h"
+#include "runtime/expr.h"
+#include "runtime/print.h"
+#include "runtime/util.h"
+
+// This visualizer works by printing a pipeline's body as javascript, where the calls and copies are implemented by
+// calling a function `produce`. This function is implemented to record the buffers that are produced, and uses those
+// records to visualize the ranges of buffers produced. To properly visualize aliases and storage folding, writes are
+// mapped to their flat address, and then translated back to a 2D coordinate.
+// TODO: Currently, this mapping just assumes dimension 0 is x, and dimension 1 is y. This needs a more clever solution
+// to handle higher rank buffers, and to handle large buffers that need to be scaled to be visualized properly.
+
+namespace slinky {
+
+namespace {
+
+class js_printer : public node_visitor {
+public:
+  int depth = -1;
+  std::ostream& os;
+  const node_context* context;
```

Comment:
nit: when/why can this legitimately be nullptr? comment perhaps?

Reply:
This is a pretty common pattern in the codebase: you only need `node_context` to associate names with `symbol_id`s. If you don't have a `node_context`, everything still works, you just get numbered symbols instead of names in stringified outputs.